plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java

```diff
@@ -332,22 +334,23 @@ static void verifyFailedCleanly(Jenkins j, WorkflowRun run) throws Exception {
 
         assert !run.isBuilding();
 
-        if (run.getExecution() instanceof  CpsFlowExecution) {
-            Assert.assertEquals(Result.FAILURE, ((CpsFlowExecution) run.getExecution()).getResult());
+        if (run.getExecution() instanceof  CpsFlowExecution cpsFlowExecution) {
+            Assert.assertEquals(executionResult, cpsFlowExecution.getResult());
         }
 
         Assert.assertEquals(Result.FAILURE, run.getResult());
         assert !run.isBuilding();
-        // TODO verify all blocks cleanly closed out, so Block start and end nodes have same counts and FlowEndNode is last node
-        verifyCompletedCleanly(j, run);
+        verifyCompletedCleanly(j, run, verifyFlowEndNode);
     }
 
     /** Verifies all the universal post-build cleanup was done, regardless of pass/fail state. */
-    static void verifyCompletedCleanly(Jenkins j, WorkflowRun run) throws Exception {
+    static void verifyCompletedCleanly(Jenkins j, WorkflowRun run, boolean verifyFlowEndNode) throws Exception {
         // Assert that we have the appropriate flow graph entries
         FlowExecution exec = run.getExecution();
-        List<FlowNode> heads = exec.getCurrentHeads();
-        Assert.assertEquals(1, heads.size());
+        if (verifyFlowEndNode) {
```

Comment:
The real purpose of a lot of these tests is unclear to me. They assert specific behaviors, but the desired high-level effect is something else.